bindings/netstandard/ElectionGuard/ElectionGuard.ElectionSetup/KeyCeremonyMediator.cs

```diff
@@ -466,203 +556,178 @@ private async Task<bool> AlwaysRun()
-
+        var backupCount = await _backupService.CountAsync(keyCeremonyId, guardianId);
         return backupCount != CeremonyDetails.NumberOfGuardians;
     }
 
-    private async Task<bool> ShouldGuardianRunStep5()
+    public async Task<bool> ShouldGuardianRunStep5()
     {
         var keyCeremonyId = CeremonyDetails.KeyCeremonyId;
         var guardianId = UserId;
 
-        GuardianPublicKeyService guardianService = new();
-        var guardian = await guardianService.GetByIdsAsync(keyCeremonyId, guardianId);
+        var guardian = await _publicKeyService.GetByIdsAsync(keyCeremonyId, guardianId);
         if (guardian == null)
         {
             return false;
         }
 
-        GuardianBackupService backupService = new();
-        var backupCount = await backupService.CountAsync(keyCeremonyId, guardianId);
+        var backupCount = await _backupService.CountAsync(keyCeremonyId, guardianId);
+
 
-        VerificationService verificationService = new();
-        var verificationCount = await verificationService.CountAsync(keyCeremonyId, guardianId);
+        var verificationCount = await _verificationService.CountAsync(keyCeremonyId, guardianId);
 
         return backupCount == CeremonyDetails.NumberOfGuardians &&
             verificationCount != CeremonyDetails.NumberOfGuardians;
     }
 
-
     private async Task RunStep1()
     {
         var keyCeremonyId = CeremonyDetails.KeyCeremonyId;
         var currentGuardianUserName = UserId;
 
         // append guardian joined to key ceremony (db)
-        GuardianPublicKeyService service = new();
-        GuardianPublicKey data = new() { KeyCeremonyId = keyCeremonyId, GuardianId = currentGuardianUserName };
-        _ = await service.SaveAsync(data);
+
+        await _publicKeyService.UpdatePublicKeyAsync(
+            keyCeremonyId, currentGuardianUserName, null);
 
         // get guardian number
-        var guardianNumber = await GetGuardianNumberAsync(service, keyCeremonyId, currentGuardianUserName);
+        var sequenceOrder = await GetGuardianSequenceOrderAsync(
+            _publicKeyService, keyCeremonyId, currentGuardianUserName);
 
         // make guardian
         var guardian = new Guardian(
             currentGuardianUserName,
-            guardianNumber,
+            sequenceOrder,
             CeremonyDetails.NumberOfGuardians,
             CeremonyDetails.Quorum,
-             keyCeremonyId);
+            keyCeremonyId);
 
         // save guardian to local drive / yubikey
         guardian.Save(keyCeremonyId);
 
         // get public key
         var publicKey = guardian.SharePublicKey();
-
-        // append to key ceremony (db)
-        try
-        {
-            await service.UpdatePublicKeyAsync(keyCeremonyId, currentGuardianUserName, publicKey);
-        }
-        catch (ElectionGuardException)
+        // read isInBounds to ensure the handle is valid
+        if (publicKey == null || publicKey.Key == null || !publicKey.Key.IsInBounds())

```

Comment:
if the `Key` is not null, can we assume this is in bounds? Or add a check inside the public key